Mathlib/RingTheory/Polynomial/Chebyshev.lean

```diff
@@ -229,10 +234,185 @@ theorem one_sub_X_sq_mul_U_eq_pol_in_T (n : ℤ) :
+theorem C_neg (n : ℤ) : C R (-n) = C R n := by
+  induction n using Polynomial.Chebyshev.induct with
+  | zero => rfl
+  | one => show X * 2 - X = X; ring
+  | add_two n ih1 ih2 =>
+    have h₁ := C_add_two R n
+    have h₂ := C_sub_two R (-n)
+    linear_combination (norm := ring_nf) (X:R[X]) * ih1 - ih2 - h₁ + h₂
+  | neg_add_one n ih1 ih2 =>
+    have h₁ := C_add_one R n
+    have h₂ := C_sub_one R (-n)
+    linear_combination (norm := ring_nf) (X:R[X]) * ih1 - ih2 + h₁ - h₂
+
+theorem C_natAbs (n : ℤ) : C R n.natAbs = C R n := by
+  obtain h | h := Int.natAbs_eq n <;> nth_rw 2 [h]; simp
+
+theorem C_neg_two : C R (-2) = X ^ 2 - 2 := by simp [C_two]
+
+theorem C_comp_two_mul_X (n : ℤ) : (C R n).comp (2 * X) = 2 * T R n := by
+  induction n using Polynomial.Chebyshev.induct with
+  | zero => simp
+  | one => simp
+  | add_two n ih1 ih2 =>
+    simp_rw [C_add_two, T_add_two, sub_comp, mul_comp, X_comp, ih1, ih2]
+    ring
+  | neg_add_one n ih1 ih2 =>
+    simp_rw [C_sub_one, T_sub_one, sub_comp, mul_comp, X_comp, ih1, ih2]
+    ring
+
+theorem C_eq_two_mul_T_comp_half_mul_X [Invertible (2 : R)] (n : ℤ) :
+    C R n = 2 * (T R n).comp (Polynomial.C ⅟2 * X) := by
+  have := congr_arg (·.comp (Polynomial.C ⅟2 * X)) (C_comp_two_mul_X R n)
+  simp_rw [comp_assoc, mul_comp, ofNat_comp, X_comp, ← mul_assoc, ← C_eq_natCast, ← C_mul,
+    Nat.cast_ofNat, mul_invOf_self', map_one, one_mul, comp_X, map_ofNat] at this
+  assumption
+
+theorem T_eq_half_mul_C_comp_two_mul_X [Invertible (2 : R)] (n : ℤ) :
+    T R n = Polynomial.C ⅟2 * (C R n).comp (2 * X) := by
+  rw [C_comp_two_mul_X, ← mul_assoc, ← map_ofNat Polynomial.C 2, ← map_mul, invOf_mul_self',
+    map_one, one_mul]
+
+/-- `S n` is the `n`th rescaled Chebyshev polynomial of the second kind (also known as a
+Vieta–Fibonacci polynomial), given by $S_n(2x) = U_n(x)$. See
+`Polynomial.Chebyshev.S_comp_two_mul_X`. -/
+noncomputable def S : ℤ → R[X]
+  | 0 => 1
+  | 1 => X
+  | (n : ℕ) + 2 => X * S (n + 1) - S n
+  | -((n : ℕ) + 1) => X * S (-n) - S (-n + 1)
+  termination_by n => Int.natAbs n + Int.natAbs (n - 1)
+
+@[simp]
+theorem S_add_two : ∀ n, S R (n + 2) = X * S R (n + 1) - S R n
+  | (k : ℕ) => S.eq_3 R k
+  | -(k + 1 : ℕ) => by linear_combination (norm := (simp [Int.negSucc_eq]; ring_nf)) S.eq_4 R k
+
+theorem S_add_one (n : ℤ) : S R (n + 1) = X * S R n - S R (n - 1) := by
+  linear_combination (norm := ring_nf) S_add_two R (n - 1)
+
+theorem S_sub_two (n : ℤ) : S R (n - 2) = X * S R (n - 1) - S R n := by
+  linear_combination (norm := ring_nf) S_add_two R (n - 2)
+
+theorem S_sub_one (n : ℤ) : S R (n - 1) = X * S R n - S R (n + 1) := by
+  linear_combination (norm := ring_nf) S_add_two R (n - 1)
+
+theorem S_eq (n : ℤ) : S R n = X * S R (n - 1) - S R (n - 2) := by
+  linear_combination (norm := ring_nf) S_add_two R (n - 2)
+
+@[simp]
+theorem S_zero : S R 0 = 1 := rfl
+
+@[simp]
+theorem S_one : S R 1 = X := rfl
+
+@[simp]
+theorem S_neg_one : S R (-1) = 0 := by simpa using S_sub_one R 0
```

Comment:
`C_neg_one` is not marked as simp; is the difference intentional?

Reply:
Yes, it's because `C R (-1)` is not in simp normal form (by `Polynomial.Chebyshev.C_neg`) and `S R (-1)` is